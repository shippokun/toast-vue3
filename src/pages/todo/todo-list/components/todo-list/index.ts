export { default as TodoListComponent } from "./todo-list.component.vue";
