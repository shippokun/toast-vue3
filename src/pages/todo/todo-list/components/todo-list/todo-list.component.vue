<template>
  <nav class="panel">
    <p class="panel-heading">TodoList</p>
    <div class="panel-block">
      <p class="control has-icons-left">
        <input
          v-model="search"
          class="input"
          type="text"
          placeholder="Search"
        />
        <span class="icon is-left">
          <i class="fas fa-search" aria-hidden="true"></i>
        </span>
      </p>
      <router-link :to="{ path: '/todos/new' }">
        <button class="button is-link is-outlined">New Todo</button>
      </router-link>
    </div>
    <p class="panel-tabs">
      <router-link to="/todos"
        ><a :class="{ 'is-active': isAll }">All</a>
      </router-link>
      <router-link :to="{ path: '/todos', query: { complated: true } }"
        ><a :class="{ 'is-active': isComplated }">Completed</a>
      </router-link>
      <router-link :to="{ path: '/todos', query: { complated: false } }"
        ><a :class="{ 'is-active': isNoncomplated }">Noncompleted</a>
      </router-link>
    </p>
    <div v-for="todo in filterTodos" :key="todo.id">
      <router-link :to="{ path: `/todos/${todo.id}` }">
        <a class="panel-block" :class="{ 'is-active': todo.complated }">
          <span class="panel-icon">
            <i class="fas fa-book" aria-hidden="true"></i>
          </span>
          {{ todo.title }}
        </a>
      </router-link>
    </div>
    <div class="panel-block">
      <button
        class="button is-link is-outlined is-fullwidth"
        @click="resetEvent"
      >
        Reset all filters
      </button>
    </div>
  </nav>
</template>
<script lang="ts">
import { defineComponent, PropType, computed, ref } from "vue";
import { Todo } from "@/models";

interface Props {
  todos: Todo[];
  complated: boolean | null;
  reset: () => void;
}

export default defineComponent({
  name: "TodoListComponent",
  props: {
    todos: {
      type: Array as PropType<Todo[]>,
      required: true,
    },
    complated: {
      type: Boolean,
      required: false,
      default: null,
    },
    reset: {
      type: Function as PropType<Props["reset"]>,
      required: true,
    },
  },
  setup(props: Props) {
    const search = ref("");
    const isComplated = computed<boolean>(() => props.complated === true);
    const isNoncomplated = computed<boolean>(() => props.complated === false);
    const isAll = computed<boolean>(() => props.complated == null);

    const filterTodos = computed(() => {
      // 文字列検索
      const result = props.todos.filter((todo) =>
        todo.title.includes(search.value)
      );
      // 達成検索
      return props.complated === null
        ? result
        : result.filter((todo) => props.complated === todo.complated);
    });

    const resetEvent = () => {
      search.value = "";
      props.reset();
    };

    return {
      isComplated,
      isNoncomplated,
      isAll,
      search,
      filterTodos,
      resetEvent,
    };
  },
});
</script>
