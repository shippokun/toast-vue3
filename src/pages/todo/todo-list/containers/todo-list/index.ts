export { default as TodoListContainer } from "./todo-list.container.vue";
