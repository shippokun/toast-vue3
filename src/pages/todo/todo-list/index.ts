export * from "./todo-list.page";
