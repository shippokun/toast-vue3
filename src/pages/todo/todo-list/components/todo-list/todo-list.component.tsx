import { RouterLink } from "vue-router";
import { defineComponent, PropType, computed, ref } from "vue";

import { ModalComponent } from "@/components";
import { formatDate } from "@/libs";
import { Todo } from "@/models";

interface Props {
  todos: Todo[] | null;
  completed: boolean | null;
  reset: () => void;
}

// TODO: class binding
// is-active ? !todo.complated
// .text-gray: { color: gray }

export const TodoListComponent = defineComponent({
  name: "TodoListComponent",
  props: {
    todos: {
      type: Array as PropType<Props["todos"]>,
      default: null,
    },
    completed: {
      type: Boolean as PropType<Props["completed"]>,
      default: null,
    },
    reset: {
      type: Function as PropType<Props["reset"]>,
      required: true,
    },
  },
  components: { ModalComponent },
  setup(props: Props) {
    const search = ref("");
    const isCompleted = computed<boolean>(() => props.completed === true);
    const isNoncompleted = computed<boolean>(() => props.completed === false);
    const isAll = computed<boolean>(() => props.completed == null);

    const filterTodos = computed(() => {
      // 文字列検索
      const result = props.todos?.filter((todo) =>
        todo.title.includes(search.value)
      );
      // 達成検索
      return props.completed === null
        ? result
        : result?.filter((todo) => props.completed === todo.completed);
    });

    const resetEvent = () => {
      search.value = "";
      props.reset();
    };

    return () => (
      <nav class="panel is-link">
        <p class="panel-heading">TodoList</p>
        <div class="panel-block">
          <p class="control has-icons-left" style="margin-right: 0.5rem">
            <input
              v-model={search.value}
              class="input"
              type="text"
              placeholder="Search"
            />
            <span class="icon is-left">
              <i class="fas fa-search" aria-hidden="true"></i>
            </span>
          </p>
          <RouterLink to={{ path: "/todos/new" }}>
            <a class="button is-link is-outlined">New Todo</a>
          </RouterLink>
        </div>
        <p class="panel-tabs">
          <RouterLink to={{ path: "/todos" }}>
            <a class={isAll.value ? "is-active" : ""}>All</a>
          </RouterLink>
          <RouterLink to={{ path: "/todos", query: { completed: "true" } }}>
            <a class={isCompleted.value ? "is-active" : ""}>Completed</a>
          </RouterLink>
          <RouterLink to={{ path: "/todos", query: { completed: "false" } }}>
            <a class={isNoncompleted.value ? "is-active" : ""}>Noncompleted</a>
          </RouterLink>
        </p>
        {filterTodos.value?.map((todo) => (
          <RouterLink to={{ path: `/todos/${todo.id}` }}>
            <a class={"panel-block is-justify-content-space-between"}>
              <span style="display: flex; align-items: center">
                <span class="panel-icon">
                  <i class="fas fa-book" aria-hidden="true"></i>
                </span>
                {todo.title}
              </span>
              {formatDate(todo.updatedAt, "MM/dd hh:mm")}
            </a>
          </RouterLink>
        ))}
        <div class="panel-block">
          <button
            class="button is-link is-outlined is-fullwidth"
            onClick={() => resetEvent()}
          >
            Reset all filters
          </button>
        </div>
      </nav>
    );
  },
});
