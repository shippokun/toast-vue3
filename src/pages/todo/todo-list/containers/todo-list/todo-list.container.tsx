import { useRouter } from "vue-router";
import { defineComponent, PropType } from "vue";

import { useTodoStore } from "@/store";

import { TodoListComponent } from "../../components";

interface Props {
  completed: boolean | null;
}

export const TodoListContainer = defineComponent({
  name: "TodoListContainer",
  components: { TodoListComponent },
  props: {
    completed: {
      type: Boolean as PropType<Props["completed"]>,
      default: null,
    },
  },
  setup(props: Props) {
    const router = useRouter();
    const { todos, fetchAll } = useTodoStore();

    fetchAll();

    const reset = (): void => {
      router.push({ path: "/todos" }).catch(() => {
        // 強制リロード
        router.go(0);
      });
    };

    return () => (
      <TodoListComponent
        todos={todos.value}
        completed={props.completed}
        reset={reset}
      />
    );
  },
});
