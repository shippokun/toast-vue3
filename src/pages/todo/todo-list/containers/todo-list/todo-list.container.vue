<template>
  <todo-list-component :todos="todos" :completed="completed" :reset="reset" />
</template>
<script lang="ts">
import { defineComponent } from "vue";
import { useRouter } from "vue-router";

import { useTodoStore } from "@/store";
import { TodoListComponent } from "../../components";

interface Props {
  completed: boolean | null;
}

export default defineComponent({
  name: "TodoListContainer",
  components: { TodoListComponent },
  props: {
    completed: {
      type: Boolean,
      required: false,
      default: null,
    },
  },
  setup(props: Props) {
    const router = useRouter();
    const { todos } = useTodoStore();

    const reset = (): void => {
      router.push({ path: "/todos" }).catch(() => {
        // 強制リロード
        router.go(0);
      });
    };

    return { todos, props, reset };
  },
});
</script>
