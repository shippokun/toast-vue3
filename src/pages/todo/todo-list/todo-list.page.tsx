import { LocationQuery, useRoute } from "vue-router";
import { defineComponent, ref, watch } from "vue";

import { TodoListContainer } from "./containers";

const toCompleted = (query: LocationQuery): boolean | null => {
  let result: boolean | null;
  switch (query["completed"]) {
    case "true":
      result = true;
      break;
    case "false":
      result = false;
      break;
    default:
      result = null;
  }
  return result;
};

export const TodoListPage = defineComponent({
  name: "TodoListPage",
  components: { TodoListContainer },
  setup() {
    const route = useRoute();
    const completed = ref(toCompleted(route.query));

    watch(
      () => route.query,
      (query) => {
        completed.value = toCompleted(query);
      }
    );

    return () => <TodoListContainer completed={completed.value} />;
  },
});
