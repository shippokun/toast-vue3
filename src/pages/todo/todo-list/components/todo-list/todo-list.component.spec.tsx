import { mount } from "@vue/test-utils";
import { createTodosMock } from "@/models/testing";
import { TodoListComponent } from "./todo-list.component";
import { router } from "@/router";

describe("todo-list.component.tsx", () => {
  describe("renders", () => {
    it("default render", () => {
      const wrapper = mount(TodoListComponent, {
        props: {
          reset: jest.fn(),
        },
        global: {
          plugins: [router],
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
    it("props todos", () => {
      const todos = createTodosMock();
      const mockFn = jest.fn();
      const wrapper = mount(TodoListComponent, {
        props: {
          todos,
          reset: mockFn,
        },
        global: {
          plugins: [router],
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
      wrapper.find("button").trigger("click");
      expect(mockFn).toHaveBeenCalled();
    });
  });
});
