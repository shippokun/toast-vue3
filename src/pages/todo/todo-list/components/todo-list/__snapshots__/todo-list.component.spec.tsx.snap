// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-list.component.tsx renders default render 1`] = `
<nav class="panel is-link">
  <p class="panel-heading">TodoList</p>
  <div class="panel-block">
    <p class="control has-icons-left" style="margin-right: 0.5rem;"><input class="input" type="text" placeholder="Search"><span class="icon is-left"><i class="fas fa-search" aria-hidden="true"></i></span></p><a href="/todos/new" class=""><a class="button is-link is-outlined">New Todo</a></a>
  </div>
  <p class="panel-tabs"><a href="/todos" class=""><a class="is-active">All</a></a><a href="/todos?completed=true" class=""><a class="">Completed</a></a><a href="/todos?completed=false" class=""><a class="">Noncompleted</a></a></p>
  <!---->
  <div class="panel-block"><button class="button is-link is-outlined is-fullwidth">Reset all filters</button></div>
</nav>
`;

exports[`todo-list.component.tsx renders props todos 1`] = `
<nav class="panel is-link">
  <p class="panel-heading">TodoList</p>
  <div class="panel-block">
    <p class="control has-icons-left" style="margin-right: 0.5rem;"><input class="input" type="text" placeholder="Search"><span class="icon is-left"><i class="fas fa-search" aria-hidden="true"></i></span></p><a href="/todos/new" class=""><a class="button is-link is-outlined">New Todo</a></a>
  </div>
  <p class="panel-tabs"><a href="/todos" class=""><a class="is-active">All</a></a><a href="/todos?completed=true" class=""><a class="">Completed</a></a><a href="/todos?completed=false" class=""><a class="">Noncompleted</a></a></p><a href="/todos/1" class=""><a class="panel-block is-justify-content-space-between"><span style="display: flex; align-items: center;"><span class="panel-icon"><i class="fas fa-book" aria-hidden="true"></i></span>title1</span>01/02 07:17</a></a><a href="/todos/2" class=""><a class="panel-block is-justify-content-space-between"><span style="display: flex; align-items: center;"><span class="panel-icon"><i class="fas fa-book" aria-hidden="true"></i></span>title2</span>01/02 07:17</a></a><a href="/todos/3" class=""><a class="panel-block is-justify-content-space-between"><span style="display: flex; align-items: center;"><span class="panel-icon"><i class="fas fa-book" aria-hidden="true"></i></span>title3</span>01/02 07:17</a></a>
  <div class="panel-block"><button class="button is-link is-outlined is-fullwidth">Reset all filters</button></div>
</nav>
`;
